File: adventofcode/2023/day01/day01.hs
import Data.Char (digitToInt, isDigit)
import Data.List (isPrefixOf)
import Debug.Trace (trace)

firstDigit :: String -> Char
firstDigit [] = '0'
firstDigit (x : xs)
    | isDigit x = x
    | otherwise = firstDigit xs

lastDigit :: String -> Char
lastDigit [] = '0'
lastDigit xs
    | isDigit x = x
    | otherwise = lastDigit $ init xs
  where
    x = last xs

parseNum :: String -> String
parseNum xs = firstDigit xs : lastDigit xs : []

part1 :: [String] -> Int
part1 = sum . map (read . parseNum)

one = "one"
two = "two"
three = "three"
four = "four"
five = "five"
six = "six"
seven = "seven"
eight = "eight"
nine = "nine"

firstDigit2 :: String -> Int
firstDigit2 [] = 0
firstDigit2 xx@(x : xs)
    | isPrefixOf one xx = 1
    | isPrefixOf two xx = 2
    | isPrefixOf three xx = 3
    | isPrefixOf four xx = 4
    | isPrefixOf five xx = 5
    | isPrefixOf six xx = 6
    | isPrefixOf seven xx = 7
    | isPrefixOf eight xx = 8
    | isPrefixOf nine xx = 9
    | isDigit x = digitToInt x
    | otherwise = firstDigit2 xs

lastDigit2 :: String -> Int
lastDigit2 [] = 0
lastDigit2 xs
    | isPrefixOf (reverse one) rs = 1
    | isPrefixOf (reverse two) rs = 2
    | isPrefixOf (reverse three) rs = 3
    | isPrefixOf (reverse four) rs = 4
    | isPrefixOf (reverse five) rs = 5
    | isPrefixOf (reverse six) rs = 6
    | isPrefixOf (reverse seven) rs = 7
    | isPrefixOf (reverse eight) rs = 8
    | isPrefixOf (reverse nine) rs = 9
    | isDigit x = digitToInt x
    | otherwise = lastDigit2 $ init xs
  where
    rs = reverse xs
    x = last xs

parseNum2 :: String -> Int
parseNum2 xs = fn * 10 + ln
  where
    fn = firstDigit2 xs
    ln = lastDigit2 xs

part2 :: [String] -> Int
part2 = sum . map parseNum2

-- main :: IO ()
main = interact $ unlines . map show . sequence [part1, part2] . lines

File: adventofcode/2023/day01/day01.zig
const std = @import("std");
const time = std.time;
const print = std.debug.print;
const toDigit = std.fmt.charToDigit;

fn first_digit(ns: []u8) u8 {
    for (ns) |n| {
        if (std.ascii.isDigit(n)) return toDigit(n, 10) catch 0;
    }
    return 0;
}

fn last_digit(ns: []u8) u8 {
    for (0..ns.len) |i| {
        if (std.ascii.isDigit(ns[ns.len - i - 1])) return toDigit(ns[ns.len - i - 1], 10) catch 0;
    }
    return 0;
}

const one = "one";
const two = "two";
const three = "three";
const four = "four";
const five = "five";
const six = "six";
const seven = "seven";
const eight = "eight";
const nine = "nine";

fn first_digit2(ns: []u8) u8 {
    const l = ns.len;
    for (0..l) |i| {
        if (i + 3 < l and std.mem.eql(u8, one, ns[i .. i + 3])) return 1;
        if (i + 3 < l and std.mem.eql(u8, two, ns[i .. i + 3])) return 2;
        if (i + 5 < l and std.mem.eql(u8, three, ns[i .. i + 5])) return 3;
        if (i + 4 < l and std.mem.eql(u8, four, ns[i .. i + 4])) return 4;
        if (i + 4 < l and std.mem.eql(u8, five, ns[i .. i + 4])) return 5;
        if (i + 3 < l and std.mem.eql(u8, six, ns[i .. i + 3])) return 6;
        if (i + 5 < l and std.mem.eql(u8, seven, ns[i .. i + 5])) return 7;
        if (i + 5 < l and std.mem.eql(u8, eight, ns[i .. i + 5])) return 8;
        if (i + 4 < l and std.mem.eql(u8, nine, ns[i .. i + 4])) return 9;
        if (std.ascii.isDigit(ns[i])) return toDigit(ns[i], 10) catch 0;
    }
    return 0;
}

fn last_digit2(ns: []u8) u8 {
    var i: usize = ns.len;
    while (i > 0) {
        i -= 1;
        if (i >= 2 and std.mem.eql(u8, one, ns[(i - 2)..(i + 1)])) return 1;
        if (i >= 2 and std.mem.eql(u8, two, ns[(i - 2)..(i + 1)])) return 2;
        if (i >= 4 and std.mem.eql(u8, three, ns[(i - 4)..(i + 1)])) return 3;
        if (i >= 3 and std.mem.eql(u8, four, ns[(i - 3)..(i + 1)])) return 4;
        if (i >= 3 and std.mem.eql(u8, five, ns[(i - 3)..(i + 1)])) return 5;
        if (i >= 2 and std.mem.eql(u8, six, ns[(i - 2)..(i + 1)])) return 6;
        if (i >= 4 and std.mem.eql(u8, seven, ns[(i - 4)..(i + 1)])) return 7;
        if (i >= 4 and std.mem.eql(u8, eight, ns[(i - 4)..(i + 1)])) return 8;
        if (i >= 3 and std.mem.eql(u8, nine, ns[(i - 3)..(i + 1)])) return 9;
        if (std.ascii.isDigit(ns[i])) return toDigit(ns[i], 10) catch 0;
    }
    return 0;
}

pub fn main() !void {
    const s = time.microTimestamp();
    const stdin = std.io.getStdIn().reader();
    var p1: u16 = 0;
    var p2: u16 = 0;
    while (true) {
        var buf: [100]u8 = undefined;
        const line = try stdin.readUntilDelimiterOrEof(&buf, '\n');
        if (line == null) break;

        const l = line.?;
        p1 += 10 * first_digit(l) + last_digit(l);
        p2 += 10 * first_digit2(l) + last_digit2(l);
    }
    print("{d}\n", .{p1});
    print("{d}\n", .{p2});
    const e = time.microTimestamp();
    print("Time taken: {s}\n", .{std.fmt.fmtDurationSigned(e - s)});
}
